grabbag of subtyping punchlist items

vtjnash opened the pull request:
From the list in https://github.com/JuliaLang/julia/issues/19998. Will check those off when this is merged.

vtjnash committed:
allow running tests in any random subdirectory, not just linalg

vtjnash committed:
fix offsetof(Task, ctx)

...by removing it.
This type can't be represented in Julia correctly,
but it also isn't needed.

vtjnash committed:
simplify / improve jl_typemap_intersection visitor now that isa should be correct

vtjnash committed:
simplify process of printing UnionAll type-variables

Since UnionAll bounds are encountered recursively,
it's easiest to represent this the same way in the IOContext.

Also avoid the performance / allocation overhead of using
`IOContext(io)` as a no-op convert (not a no-op since keyword support was added)

vtjnash committed:
correct the ambiguity detection test in ml_matches

the previous version was effectively equivalent, but did more work than necessary
and could fail in the wrong direction in the unlikely case that `ambi` was not a strict subtype of `mambig->sig`.

vtjnash committed:
remove unnecessary work from check_ambiguous_visitor

vtjnash committed:
add parenthetical to manual regarding absence of Float typealias

Seems to be a common FAQ. Not sure if those users would have seen it here,
but it's right after explaining the existance of the `Int` type-alias.

vtjnash committed:
add supertype operator >: for symmetry with issubtype

vtjnash committed:
fix typetree dump to consider that a UnionAll may be the primary type declaration

vtjnash committed:
fix typetree example to demonstrate new type representations and subtyping relationships

tkelman commented on doc/src/manual/types.md:
```diff
@@ -1045,6 +1045,10 @@ end
 type -- either `Int32` or `Int64`.
 
+(Note that unlike `Int`, `Float` does not exist as a type-alias for a specific sized `AbstractFloat`.
+Unlike with integer registers, the size of the floating point representation is problem dependent,
```
Integer sizes often need to be problem dependent too, I'm not sure this is the right wording to justify this.

vtjnash replied:
Hm. Maybe I should say something along the lines of "the floating point register sizes are standardized across CPUs based on the IEEE-754 standard."?

tkelman replied:
practically speaking I'd say that's more accurately the reason

tkelman commented on doc/src/stdlib/punctuation.md:
```diff
@@ -42,3 +42,6 @@ Extended documentation for mathematical symbols & functions is [here](@ref math-
 | `:a`        | symbol a                                                                                    |
+| `<:`        | [`subtype operator`](@ref <:)                                                               |
+| `>:`        | [`supertype operator`](@ref >:) (reverse of subtype operator)                               |
+| `===`       | [`eqal comparison operator`](@ref ===)                                                      |
```
eqal or egal?

tkelman commented on base/show.jl:
```diff
@@ -19,24 +19,28 @@ immutable IOContext{IO_t <: IO} <: AbstractPipe
+#The same as `IOContext(io::IO, KV::Pair)`, but accepting properties as keyword arguments.
+#"""
+#IOContext(io::IO; kws...) = IOContext(IOContext(io); kws...)
+#function IOContext(io::IOContext; kws...)
```
this might break Documenter, Juno, and IJulia, is it possible to deprecate this?

vtjnash replied:
Not easily. The problem is that it replaces the primitive `IOContext(io)` constructor (which would otherwise be inlined), with several kwarg calls. (since the kwarg method needs to be the same as the no-keyword args function)

JeffBezanson replied:
So the commit message should say "remove keyword argument `IOContext` constructors". That's a pretty sneaky thing to combine with an implementation-only change.

StefanKarpinski replied:
Seriously – this is not non-breaking...

vtjnash replied:
you mean sneaky like the commit message that added the alternate constructor? "use IOContext more consistently" (ce231746d)

JeffBezanson replied:
Unfortunately there's a ratchet effect. Adding a feature or method doesn't break anything, but removing one does.

JeffBezanson replied:
Also, why does the performance of this constructor matter?

JeffBezanson commented on base/operators.jl:
```diff
@@ -9,11 +9,18 @@ typealias Indices{N} NTuple{N,AbstractUnitRange}
+
+Supertype operator, equivalent to `issubtype(T2, T1)`.
+"""
+const (>:)(a::ANY, b::ANY) = issubtype(b, a)
```
I believe this needs special inference support to be able to infer `Const(true)` or `Const(false)`.

vtjnash replied:
Probably doable, and perhaps not a bad idea anyways (so that it doesn't attempt to actually specialize on any arguments that it sees). I guess the idea is that it's preferable to ensure `>:` and `<:` behave the same. Should I make both of them (`issupertype`) builtins to keep it identical, or just make it close enough via inference?

JeffBezanson replied:
I'd rather not add more builtins; let's just add it to inference.

vtjnash committed:
remove several unnecessary special-case inliners

the general inliner is better and also respects effect_free

(related to #9765, although none of these specific cases were mentioned there)

vtjnash commented:
bump

JeffBezanson commented on test/show.jl:
```diff
@@ -569,7 +569,7 @@ let repr = sprint(dump, Any)
     @test ismatch(r"^Any\n  [^ \t\n]", repr)
     @test endswith(repr, '\n')
-    @test_broken contains(repr, "     Base.Vector{T} = Array{T,1}\n")
+    @test contains(repr, "     Base.Vector{T} = Array{T,1} where T\n")
```
It would be more accurate print either `Base.Vector = Array{T,1} where T` or `Base.Vector{T} = Array{T,1}`. That second syntax will be added soon.

vtjnash replied:
done

vtjnash committed:
use new "where" syntax in dumpsubtypes

JeffBezanson commented:
LGTM.

KristofferC commented:
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/20407#issuecomment-278776409) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/d75e9ce3df72cada6b230b58eb931514768988c6/7dee3b9_vs_ddb1e1e/report.md). cc @jrevels

vtjnash committed:
Merge branch 'master' into jn/grabbag-subtypes

tkelman commented:
the comprehension slowdown there looks like it could be real
